vsintegration/src/FSharp.Editor/CodeFixes/MakeOuterBindingRecursive.fs

```diff
@@ -9,45 +9,38 @@ open System.Collections.Immutable
 open Microsoft.CodeAnalysis.Text
 open Microsoft.CodeAnalysis.CodeFixes
 
+open CancellableTasks
+
 [<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = CodeFix.MakeOuterBindingRecursive); Shared>]
 type internal MakeOuterBindingRecursiveCodeFixProvider [<ImportingConstructor>] () =
     inherit CodeFixProvider()
 
     override _.FixableDiagnosticIds = ImmutableArray.Create("FS0039")
 
-    override _.RegisterCodeFixesAsync context =
-        asyncMaybe {
-            let! parseResults =
-                context.Document.GetFSharpParseResultsAsync(nameof (MakeOuterBindingRecursiveCodeFixProvider))
-                |> liftAsync
-
-            let! sourceText = context.Document.GetTextAsync(context.CancellationToken)
-
-            let diagnosticRange =
-                RoslynHelpers.TextSpanToFSharpRange(context.Document.FilePath, context.Span, sourceText)
-
-            do! Option.guard (parseResults.IsPosContainedInApplication diagnosticRange.Start)
-
-            let! outerBindingRange = parseResults.TryRangeOfNameOfNearestOuterBindingContainingPos diagnosticRange.Start
-            let! outerBindingNameSpan = RoslynHelpers.TryFSharpRangeToTextSpan(sourceText, outerBindingRange)
-
-            // One last check to verify the names are the same
-            do!
-                Option.guard (
-                    sourceText
-                        .GetSubText(outerBindingNameSpan)
-                        .ContentEquals(sourceText.GetSubText(context.Span))
-                )
-
-            let title =
-                String.Format(SR.MakeOuterBindingRecursive(), sourceText.GetSubText(outerBindingNameSpan).ToString())
-
-            do
-                context.RegisterFsharpFix(
-                    CodeFix.MakeOuterBindingRecursive,
-                    title,
-                    [| TextChange(TextSpan(outerBindingNameSpan.Start, 0), "rec ") |]
-                )
-        }
-        |> Async.Ignore
-        |> RoslynHelpers.StartAsyncUnitAsTask(context.CancellationToken)
+    override this.RegisterCodeFixesAsync context = context.RegisterFsharpFix(this)
+
+    interface IFSharpCodeFixProvider with
+        member _.GetCodeFixIfAppliesAsync context =
+            cancellableTask {
+                let! parseResults = context.GetParseResultsAsync(nameof MakeOuterBindingRecursiveCodeFixProvider)
+                let! sourceText = context.GetSourceTextAsync()
+                let! diagnosticRange = context.GetErrorRangeAsync()
+
+                return
+                    Option.guard (parseResults.IsPosContainedInApplication diagnosticRange.Start)
+                    |> Option.bind (fun () -> parseResults.TryRangeOfNameOfNearestOuterBindingContainingPos diagnosticRange.Start)
+                    |> Option.bind (fun bindingRange -> RoslynHelpers.TryFSharpRangeToTextSpan(sourceText, bindingRange))
+                    |> Option.filter (fun bindingSpan ->
```

Comment:
`filter + map` is just a bind with an explicitely admitted `if`.
Similar to the above - it did make sense together with asyncMaybe DSL.
Without it, it's way more readable to use plain `if` statements where there are clear boolean decisions.

Reply:
Same here - personal preference, we should have a code convention for these. `Option.filter` is a part of the standard library, to avoid the ladder-looking code. I'd prefer to keep it.

Reply:
I also prefer the pipeline style to explicit `if`s

Reply:
TBH I think this repo is quite clear on the usage of plain 'if'/'match' vs. using Option.filter , I think this convention stands out quite implicitly.